HEN_HOUSE/egs++/geometry/egs_cd_geometry/egs_cd_geometry.h

```diff
@@ -506,9 +506,25 @@ class EGS_CDGEOMETRY_EXPORT EGS_CDGeometry : public EGS_BaseGeometry {
                     }
                     return ixold;
                 }
-                else if (ixnew < 0 && tb <= epsilon) {                         // (a) is true
+				// Skip for concave geometries: particles can reenter the CD geometry after leaving.
+				// Check 2. below will catch these cases.
+                else if ((ixnew < 0 && tb <= epsilon) && (bg->isConvex() && g[ibase]->isConvex())) {                         // (a) is true
                     return ixnew;
                 }
+				// If a particle approaching the geometry sits on a boundary, we look back to see
+				// if we just entered the geometry (the previous checks fail to catch this case).
+				EGS_Float tb_neg = 1e30;
+				int ixnew_neg = howfar(ixold,x,u*(-1),tb_neg,0,pn);
+				if (ixnew_neg < 0 && tb_neg <= epsilon) {                             // (b) is true
+					t = 0;
+					if (newmed) *newmed = g[ibase] ? g[ibase]->medium(icd) :
+						bg->medium(ibase);
+					if( normal ) {
+						*normal = (*pn)*(-1);
+					}
+					return ixold;
+				}
+
                 //***********************************************************************
```

Comment:
Notice that this is similar to what the code previously used to check for condition (b), but it wasn't enough to catch everything. It turns out the current check for (b) is also not enough to cover all cases. Hopefully, this improves the coverage of edge cases.